src/cargo/core/compiler/output_sbom.rs

```diff
@@ -0,0 +1,292 @@
+
+        Self {
+            format_version: SbomFormatVersion,
+            package_id,
+            name,
+            version,
+            source,
+            target,
+            profile,
+            packages,
+            features,
+            rustc,
+        }
+    }
+}
+
+/// Saves a `<artifact>.cargo-sbom.json` file for the given [`Unit`].
+///
+pub fn output_sbom(build_runner: &mut BuildRunner<'_, '_>, unit: &Unit) -> CargoResult<()> {
+    let bcx = build_runner.bcx;
+    let rustc: SbomRustc = bcx.rustc().into();
+
+    let packages = collect_packages(build_runner, unit);
+
+    for sbom_output_file in build_runner.sbom_output_files(unit)? {
+        let sbom = Sbom::new(unit, packages.clone(), rustc.clone());
+
+        let outfile = BufWriter::new(paths::create(sbom_output_file)?);
+        serde_json::to_writer(outfile, &sbom)?;
+    }
+
+    Ok(())
+}
+
+/// Fetch all dependencies, including transitive ones. A dependency can also appear multiple times
+/// if it's using different settings, e.g. profile, features or crate versions.
+///
+/// A package's `dependencies` is a list of indices into the `packages` array.
+fn collect_packages(build_runner: &mut BuildRunner<'_, '_>, unit: &Unit) -> Vec<SbomPackage> {
+    let unit_graph = &build_runner.bcx.unit_graph;
+    let root_deps = build_runner.unit_deps(unit);
+    let root_profile = &unit.profile;
+
+    let mut packages = Vec::new();
+    let mut queue: BTreeSet<&UnitDep> = root_deps.iter().collect();
+    let mut visited_dependencies = BTreeSet::new();
+
+    // each package should appear only once in the `packages` block
+    while let Some(dependency) = queue.pop_first() {
+        if visited_dependencies.contains(dependency) {
+            continue;
+        }
+
+        let mut dependencies: BTreeSet<&UnitDep> = unit_graph[&dependency.unit].iter().collect();
+
+        // each seen package is returned
+        packages.push(SbomPackage::new(dependency, root_profile));
+
+        // append dependencies to back of queue
+        queue.append(&mut dependencies);
+        visited_dependencies.insert(dependency);
+    }
+
+    // Collect all indices for each package's dependencies
+    //
+    // At this point `visited_packages` contains the [`UnitDep`] the `packages` Vec is generated with.
+    // They have the same number of objects in the same order, therefore using an index is fine.
+    for (index, package) in visited_dependencies.iter().enumerate() {
+        let dependencies: BTreeSet<&UnitDep> = unit_graph[&package.unit].iter().collect();
+
+        let mut indices = dependencies
+            .iter()
+            .filter_map(|dep| {
+                visited_dependencies
+                    .iter()
+                    .position(|unit_dep| dep == unit_dep)
+            })
+            .collect::<Vec<_>>();
+
+        if let Some(package) = packages.get_mut(index) {
+            package.dependencies.append(&mut indices);
```

Comment:
This assumes that the order we iterate through `visitied_dependencies` (`BTreeSet`) is the same order things appear in `packages (`Vec`).

Should `visited_dependencies` be an `IndexSet`?